betterproto/__init__.py

```diff
@@ -426,6 +430,81 @@ def parse_fields(value: bytes) -> Generator[ParsedField, None, None]:
 T = TypeVar("T", bound="Message")
 
 
+class ProtoClassMetadata:
+    cls: "Message"
+
+    def __init__(self, cls: "Message"):
+        self.cls = cls
+        by_field = {}
+        by_group = {}
+
+        for field in dataclasses.fields(cls):
+            meta = FieldMetadata.get(field)
+
+            if meta.group:
+                # This is part of a one-of group.
+                by_field[field.name] = meta.group
+
+                by_group.setdefault(meta.group, set()).add(field)
+
+        self.oneof_group_by_field = by_field
+        self.oneof_field_by_group = by_group
+
+    def init_default_gen(self):
+        default_gen = {}
+
+        for field in dataclasses.fields(self.cls):
+            meta = FieldMetadata.get(field)
+            default_gen[field.name] = self.cls._get_field_default_gen(field, meta)
+
+        self.default_gen = default_gen
+
+    def init_cls_by_field(self):
+        field_cls = {}
+
+        for field in dataclasses.fields(self.cls):
+            meta = FieldMetadata.get(field)
+            if meta.proto_type == TYPE_MAP:
+                assert meta.map_types
+                kt = self.cls._cls_for(field, index=0)
+                vt = self.cls._cls_for(field, index=1)
+                Entry = dataclasses.make_dataclass(
+                    "Entry",
+                    [
+                        ("key", kt, dataclass_field(1, meta.map_types[0])),
+                        ("value", vt, dataclass_field(2, meta.map_types[1])),
+                    ],
+                    bases=(Message,),
+                )
+                make_protoclass(Entry)
+                field_cls[field.name] = Entry
+                field_cls[field.name + ".value"] = vt
+            else:
+                field_cls[field.name] = self.cls._cls_for(field)
+
+        self.cls_by_field = field_cls
+
+    def __getattr__(self, item):
+        # Lazy init because forward reference classes may not be available at the beginning.
+        if item == 'default_gen':
+            self.init_default_gen()
+            return self.default_gen
+
+        if item == 'cls_by_field':
+            self.init_cls_by_field()
+            return self.cls_by_field
+
+
+def make_protoclass(cls):
+    setattr(cls, "_betterproto", ProtoClassMetadata(cls))
```

Comment:
Maybe this could be achieved in a backwards compatible (and maybe more direct?) way by making ProtoClassMetadata a [non-data descriptor](https://realpython.com/python-descriptors/) of the Message class?

So something like: 

```python
class ProtoClassMetadata:
    def __lazy_init__(self, cls: "Message"):  # rename __init__ to __lazy_init__
        ...
    
    def __get__(self, obj, type=None):
        if self.cls is None:
            self.__lazy_init__(obj.__class__)
        return self

class Message:
    _betterproto = ProtoClassMetadata()
    ...
```

Reply:
Good idea! But this will add a class field to class `Message`, so that all subclasses share one instance of `ProtoClassMetadata`.

The following approach works:

```python
class ProtoMessageMetadata:
    metadata: Dict[Type["Message"], ProtoClassMetadata]

    def __init__(self):
        self.metadata = {}

    def __get__(self, instance, owner):
        meta = self.metadata.get(owner)
        if not meta:
            meta = ProtoClassMetadata(owner)
            self.metadata[owner] = meta
        return meta
```

which stores metadata of all subclasses in `Message` class. The drawback is that it references all subclasses and those dynamically created classes won't be released.

Let me see if we could use `@property` to solve this.

Reply:
It works. Now it's transparent to the users. Thanks!

Reply:
Nice! Except the memory leak :/ What about using a weakref data structure to avoid the data leak scenario?

Or maybe it’s simpler if the reference to the metadata is actually stored in the target class itself... so there’s no need to keep the reference to the class. The metadata could still be accessed through this ProtoMessageMetadata.__get__, or through a regular @property on message that lazy initialises it.